### PR TITLE
SNOW-621669 ml transformer: Normalizer and KBinsDiscretizer

### PR DESCRIPTION
**Pull request by sfc-gh-yuwang**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes #NNNN

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/ml/transformer.py [hunk 1]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+                    snow_pow(
+                        snow_sum(snow_pow(col(input_col), norm_p)), 1 / norm_p
+                    ).as_(f"{input_col}_l{norm_p}_norm")
+                    for input_col in self.input_cols
+                ]
+            ).write.save_as_table(self._norm_table_name, create_temp_table=True)
+        else:
+            df.select(
+                [
+                    snow_max(col(input_col)).as_(f"{input_col}_max_norm")
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:615]: inf norm should use the maximum of absolute value

File: src/snowflake/snowpark/ml/transformer.py [hunk 2]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+        self._fitted = False
+
+    def fit(self, df: DataFrame) -> "LabelEncoder":
+        super().fit(df)
+        self._fitted = True
+        category_col_name = df.columns[0]
+        df_label = (
+            df._distinct([category_col_name])
+            .select([col(category_col_name).as_("label_name")])
+            .with_column("label_index", builtin("seq8")())
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:530]: You should also use self._states_col

File: src/snowflake/snowpark/ml/transformer.py [hunk 3]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+        ):
+            results.append(output_col)
+            column_res.append(
+                iff(
+                    df[input_col] == df[states_col]["max"],
+                    self.n_bins,
+                    builtin("width_bucket")(
+                        df[input_col],
+                        df[states_col]["min"],
+                        df_state_table[states_col]["max"],
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:776]:
```suggestion
                        df[states_col]["max"],
```
?

File: src/snowflake/snowpark/ml/transformer.py [hunk 4]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+            |4.0  |
+            -------
+            """
+            for i in range(self.n_bins):
+                df_temp = df.select(
+                    [
+                        lit(i + 1).as_(output_col),
+                        iff(
+                            lit(i) == lit(0),
+                            builtin("percentile_disc")(0).within_group(input_col) - 1,
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:714]: I suggest not using -1 or +1, but use `lit(-sys.maxsize)` and `lit(sys.maxsize)` directly

File: src/snowflake/snowpark/ml/transformer.py [hunk 5]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+        """
+        ln norm is computed in this way:
+        (sum(x^n))^(1/n)
+        """
+        if self._norm[0] == "l":
+            norm_p = int(self._norm[1:])
+            df.select(
+                [
+                    snow_pow(
+                        snow_sum(snow_pow(col(input_col), norm_p)), 1 / norm_p
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:607]: This is not true for l1 norm, which is the summation of all absolute values

File: src/snowflake/snowpark/ml/transformer.py [hunk 6]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+        norm: Optional[str] = "l2",
+    ) -> None:
+        super().__init__(input_cols, output_cols)
+        self._norm = norm
+        self._norm_table_name = random_name_for_temp_object(TempObjectType.TABLE)
+
+    def fit(self, df: DataFrame) -> "Normalizer":
+        super().fit(df)
+        if self._norm != "max" and (
+            self._norm[0] != "l" or not self._norm[1:].isdigit()
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:595]: there is no l0 norm and should raise exception

File: tests/integ/ml/test_transformer.py [hunk 7]
```diff
@@ -548,3 +551,144 @@ def test_binarizer(session):
     Utils.check_answer(
         res2, [Row(A=0, B=1, C=1), Row(A=1, B=0, C=1), Row(A=0, B=None, C=None)]
     )
+
+
+def test_normalizer(session):
```
Comment (sfc-gh-jdu) [inline tests/integ/ml/test_transformer.py:556]: l1 norm and probably add l3/4 norm, and add negative test cases

File: tests/integ/ml/test_transformer.py [hunk 8]
```diff
@@ -548,3 +551,144 @@ def test_binarizer(session):
...
+    df = session.create_dataframe(
+        [[1, 4, 5], [2, 4, 5], [5, 3, 6]], schema=["a", "b", "c"]
```
Comment (sfc-gh-jdu) [inline tests/integ/ml/test_transformer.py:558]: try negative number, nan (and probably also try NULL)

File: tests/integ/ml/test_transformer.py [hunk 9]
```diff
@@ -548,3 +551,144 @@ def test_binarizer(session):
...
+    df_unkown = session.create_dataframe(
+        [[1, 4, 5], [2, 4, 5], [5, 7, 6]], schema=["a", "b", "c"]
+    )
+    df_inverse = session.create_dataframe([[0], [2], [1]], schema=["a"])
+    labelencoder = LabelEncoder()
+    labelencoder.input_cols = ["A", "B"]
+    labelencoder.output_cols = ["res_A", "res_B"]
+    model = labelencoder.fit(df_label)
+
+    res = model.transform(df)
```
Comment (sfc-gh-jdu) [inline tests/integ/ml/test_transformer.py:607]: can you use this model do inverse_transform() as well, and also unknown values for inverse_transform().

File: tests/integ/ml/test_transformer.py [hunk 10]
```diff
@@ -548,3 +551,144 @@ def test_binarizer(session):
...
+    kbins.input_cols = ["A", "B"]
+    kbins.output_cols = ["res_A", "res_B"]
+    with pytest.raises(ValueError) as ex_info:
+        kbins.fit(df)
+    assert "Wrong strategy, strategy has to be either uniform or quantile" in str(
+        ex_info.value
+    )
+
+    # negative tests: bad quantile number or dirty data
+    df = session.createDataFrame(
```
Comment (sfc-gh-jdu) [inline tests/integ/ml/test_transformer.py:686]:
```suggestion
    df = session.create_dataframe(
```

File: src/snowflake/snowpark/ml/transformer.py [hunk 11]
```diff
@@ -501,3 +508,303 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+                [
+                    count_distinct(col(f"{states_col}_lower")),
+                    count(col(f"{states_col}_lower")),
+                    count_distinct(col(f"{states_col}_upper")),
+                    count(col(f"{states_col}_upper")),
+                ]
+            )
+            judge = df_judge.collect()
+            if judge[0][0] != judge[0][1] or judge[0][2] != judge[0][3]:
+                raise ValueError("Quantile partition can not be the same")
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:742]: I suggest to use the similar error message to sklearn

File: tests/integ/ml/test_transformer.py [hunk 12]
```diff
@@ -548,3 +551,144 @@ def test_binarizer(session):
...
+    model = labelencoder.fit(df_inverse_label)
+    res = model.inverse_transform(df_inverse)
+    Utils.check_answer(
+        res, [Row(A=0, RES_A="aa"), Row(A=2, RES_A="cc"), Row(A=1, RES_A="bb")]
+    )
+
+
+def test_kbins_discretizer(session):
+    df = session.createDataFrame(
+        [[1.5, 3, 5], [2.0, 4, 7], [3.0, 5, 6], [4.0, 6, 8]], schema=["a", "b", "c"]
```
Comment (sfc-gh-jdu) [inline tests/integ/ml/test_transformer.py:641]: try nan and negative value?

File: src/snowflake/snowpark/ml/transformer.py [hunk 13]
```diff
@@ -0,0 +1,770 @@
...
+            [column_out for column_out in column_output]
+        )
+
+
+class KBinsDiscretizer(Transformer):
+    def __init__(
+        self,
+        input_cols: Optional[List[str]] = None,
+        output_cols: Optional[List[str]] = None,
+        strategy: Optional[str] = "quantile",
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:603]: the default is uniform

File: src/snowflake/snowpark/ml/transformer.py [hunk 14]
```diff
@@ -0,0 +1,770 @@
...
+            self.input_cols, self.output_cols, self._states_table_cols
+        ):
+            results.append(output_col)
+            df_quantile = session.table(self._states_table_name[input_col])
+            """
+            add columns of upper bound and lower bound of each quantile
+            """
+            df_quantile = df_quantile.with_column(
+                f"{states_col}_lower",
+                lag(df_quantile[states_col], +1, -(2**53)).over(
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:739]:
```suggestion
                lag(df_quantile[states_col], 1, -(2**53)).over(
```
why 2 ** 53?

File: src/snowflake/snowpark/ml/transformer.py [hunk 15]
```diff
@@ -0,0 +1,770 @@
...
+                    Window.order_by(df_quantile[states_col])
+                ),
+            )
+            df_quantile = df_quantile.with_column(
+                f"{states_col}_upper", df_quantile[states_col]
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:744]: You don't need a new column, right?

**Review by sfc-gh-jdu — APPROVED** [review]

File: src/snowflake/snowpark/ml/transformer.py [hunk 16]
```diff
@@ -501,3 +512,257 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+            norm_p = int(self._norm[1:])
+            df.select(
+                [
+                    pow_(
+                        sum_(
+                            pow_(
+                                abs_(
+                                    col(input_col)
+                                    if norm_p % 2 == 1
+                                    else col(input_col)
```
[the remark this answers is not in the document]
Comment (sfc-gh-yuwang) [reply src/snowflake/snowpark/ml/transformer.py:564]: sorry, this part should be 
abs_(
                                    col(input_col)
                                ) if norm_p % 2 == 1
                                else col(input_col)
i'll amend it.

File: src/snowflake/snowpark/ml/transformer.py [hunk 17]
```diff
@@ -501,3 +512,257 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+            self._states_table_name, create_temp_table=True, mode="overwrite"
+        )
+        return self
+
+    def _quantile_fit(self, df: DataFrame) -> "KBinsDiscretizer":
+        self._states_table_name = {}
+        for input_col, output_col, states_col in zip(
+            self.input_cols, self.output_cols, self._states_table_cols
+        ):
+            df_quantile = None
```
[the remark this answers is not in the document]
Comment (sfc-gh-yuwang) [reply src/snowflake/snowpark/ml/transformer.py:652]: A empty DF may make the final sql more complicated? I think it's ok to keep it this way.

File: src/snowflake/snowpark/ml/transformer.py [hunk 18]
```diff
@@ -501,3 +512,252 @@ def transform(self, df: DataFrame) -> DataFrame:
...
+        self._input_cols = value
+        self._states_table_cols = (
+            [f"states_{input_col}" for input_col in self._input_cols]
+            if self._input_cols
+            else None
+        )
+
+    def fit(self, df: DataFrame) -> "Normalizer":
+        super().fit(df)
+        if self._norm != "max" and (
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/ml/transformer.py:544]: Actually you should move this validation to `__init__`

**Pull request merged**